src/librustc_session/lint/builtin.rs

```diff
@@ -474,15 +474,34 @@ declare_lint! {
 
 declare_lint! {
     pub INDIRECT_STRUCTURAL_MATCH,
-    // defaulting to allow until rust-lang/rust#62614 is fixed.
-    Allow,
-    "pattern with const indirectly referencing non-structural-match type",
+    Warn,
+    "constant used in pattern contains value of non-structural-match type",
     @future_incompatible = FutureIncompatibleInfo {
         reference: "issue #62411 <https://github.com/rust-lang/rust/issues/62411>",
         edition: None,
     };
 }
 
+declare_lint! {
+    pub POINTER_STRUCTURAL_MATCH,
+    Allow,
+    "pointers are not structural-match",
+    @future_incompatible = FutureIncompatibleInfo {
+        reference: "issue #62411 <https://github.com/rust-lang/rust/issues/70861>",
+        edition: None,
+    };
+}
+
+declare_lint! {
+    pub NONTRIVIAL_STRUCTURAL_MATCH,
+    Warn,
+    "constant used in pattern of non-structural-match type",
```

Comment:
The lint name and description do not align. What does "non-trivial" refer to?

Reply:
I expanded on the description, is it clearer now?

Reply:
I am still somewhat confused. Why isn't it just "constant of non-structural-match type used in pattern"? What's that "triviality" condition?

Reply:
if the MIR has no conditions, function calls and basically looks like a pattern, then we accept it

Reply:
What about things like unary/binary operators or pointer derefs?
And this entirely overrides the structural match thing? How does that make any sense? Where's that analysis described in more detail?

Reply:
I feel like https://github.com/rust-lang/const-eval/blob/master/patterns.md needs a lot more content, hopefully written by someone who's already familiar with what is happening. ;)

Reply:
Hey you're as suprised about this as I am :laughing: I learned about this thing last week when I was trying to figure out things in pattern matching.

Reply:
Rereading https://github.com/rust-lang/rust/blob/7e11379f3b4c376fbb9a6c4d44f3286ccc28d149/src/librustc_mir/transform/check_consts/qualifs.rs#L116 it seems like I misjudged what's going on, so all we do is check that the mir body contains no !StructuralEq types?